File: src/main/java/view/Console.java
package view;

import model.support.allEnums.ArmorType;
import model.unit.Unit;

public abstract class Console {
    public static void print(Unit unit) {
        System.out.println("");
        System.out.println("=================");
        System.out.println("Main Parameters:");
        System.out.println("=================");
        System.out.println();
        System.out.println("Concentration: " + unit.concentration());
        System.out.println("Strength: " + unit.strength());
        System.out.println("Agility: " + unit.agility());
        System.out.println();
        System.out.println("Health: " + unit.health() + "/" + unit.maxHealth());
        System.out.println("Stamina: " + unit.stamina() + "/" + unit.maxStamina());
        System.out.println();
        System.out.println("Equipment weight: " + unit.equipmentWeight());
        System.out.println();
        System.out.println("Needed stamina for weapon attack: " + unit.neededStaminaToWeaponAttack() + "/sec");
        System.out.println("Needed stamina for hand-to-hand attack: " + unit.neededStaminaForHandToHandAttack() + "/sec");
        System.out.println();
        System.out.println("Hand-to-hand attack speed: " + unit.handToHandAttackSpeed() + "/sec");
        System.out.println("Weapon attack speed: " + unit.weaponAttackSpeed() + "/sec");

        System.out.println("");
        System.out.println("=================");
        System.out.println("Total Defense:");
        System.out.println("=================");
        System.out.println();
        unit.totalDefense().forEach((k, v) -> System.out.println(k + " = " + v));

        if (unit.weapon() != null) {
            System.out.println("");
            System.out.println("=================");
            System.out.println("Weapon:");
            System.out.println("=================");
            System.out.println();
            System.out.println("Type: " + unit.weapon().getWeaponType().type);
            System.out.println("Weight: " + unit.weapon().getWeight());
            System.out.println("Quality: " + unit.weapon().getQuality());
            System.out.println("Price: " + unit.weapon().getPrice());
            System.out.println();
            unit.weapon().getDamage().forEach((k, v) -> System.out.println(k + " = " + v));

            System.out.println("");
            System.out.println("=================");
            System.out.println("Weapon Damage:");
            System.out.println("=================");
            System.out.println();
            unit.weaponDamage().forEach((k, v) -> System.out.println(k + " = " + v + " (" + unit.criticalWeaponDamage().get(k) + ")"));
        }

        if (unit.shield() != null) {
            System.out.println("");
            System.out.println("=================");
            System.out.println("Shield:");
            System.out.println("=================");
            System.out.println();
            System.out.println("Type: " + unit.shield().getShieldType().type);
            System.out.println("Weight: " + unit.shield().getWeight());
            System.out.println("Quality: " + unit.shield().getQuality());
            System.out.println("Price: " + unit.shield().getPrice());
            System.out.println();
            unit.shield().getDefense().forEach((k, v) -> System.out.println(k + " = " + v));
        }

        if (unit.armors().containsKey(ArmorType.HELM)) {
            System.out.println("");
            System.out.println("=================");
            System.out.println("Helm:");
            System.out.println("=================");
            System.out.println();
            System.out.println("Type: " + ArmorType.HELM);
            System.out.println("Weight: " + unit.armors().get(ArmorType.HELM).getWeight());
            System.out.println("Quality: " + unit.armors().get(ArmorType.HELM).getQuality());
            System.out.println("Price: " + unit.armors().get(ArmorType.HELM).getPrice());
            System.out.println();
            unit.armors().get(ArmorType.HELM).getDefense().forEach((k, v) -> System.out.println(k + " = " + v));
        }

        if (unit.armors().containsKey(ArmorType.BODY)) {
            System.out.println("");
            System.out.println("=================");
            System.out.println("Body:");
            System.out.println("=================");
            System.out.println();
            System.out.println("Type: " + ArmorType.BODY);
            System.out.println("Weight: " + unit.armors().get(ArmorType.BODY).getWeight());
            System.out.println("Quality: " + unit.armors().get(ArmorType.BODY).getQuality());
            System.out.println("Price: " + unit.armors().get(ArmorType.BODY).getPrice());
            System.out.println();
            unit.armors().get(ArmorType.BODY).getDefense().forEach((k, v) -> System.out.println(k + " = " + v));
        }

        if (unit.armors().containsKey(ArmorType.GLOVES)) {
            System.out.println("");
            System.out.println("=================");
            System.out.println("Gloves:");
            System.out.println("=================");
            System.out.println();
            System.out.println("Type: " + ArmorType.GLOVES);
            System.out.println("Weight: " + unit.armors().get(ArmorType.GLOVES).getWeight());
            System.out.println("Quality: " + unit.armors().get(ArmorType.GLOVES).getQuality());
            System.out.println("Price: " + unit.armors().get(ArmorType.GLOVES).getPrice());
            System.out.println();
            unit.armors().get(ArmorType.GLOVES).getDefense().forEach((k, v) -> System.out.println(k + " = " + v));
        }

        if (unit.armors().containsKey(ArmorType.BOOTS)) {
            System.out.println("");
            System.out.println("=================");
            System.out.println("Boots:");
            System.out.println("=================");
            System.out.println();
            System.out.println("Type: " + ArmorType.BOOTS);
            System.out.println("Weight: " + unit.armors().get(ArmorType.BOOTS).getWeight());
            System.out.println("Quality: " + unit.armors().get(ArmorType.BOOTS).getQuality());
            System.out.println("Price: " + unit.armors().get(ArmorType.BOOTS).getPrice());
            System.out.println();
            unit.armors().get(ArmorType.BOOTS).getDefense().forEach((k, v) -> System.out.println(k + " = " + v));
        }

        System.out.println("");
        System.out.println("=================");
        System.out.println("Hand-to-hand Damage:");
        System.out.println("=================");
        System.out.println();
        System.out.println(unit.handToHandDamage() + " (" + unit.criticalHandToHandDamage() + ")");

//        System.out.println("");
//        System.out.println("=================");
//        System.out.println("Skills:");
//        System.out.println("=================");
//        System.out.println();
//        unit.skills().forEach((k, v) -> System.out.println(k + " = " + v));
    }
}

File: src/main/java/view/Console.kt
package view

import model.Character
import model.equipment.armor.Armor
import model.support.Constants.MAX
import model.support.enums.ArmorType.*
import model.support.enums.BodyPart.*
import model.support.enums.Skill
import model.support.enums.Specification

object Console {
    fun print(character: Character) {
        println()
        println("=================")
        println("Character ${character.who}")
        println("=================")
        println()
        println("Is alive: ${character.isAlive}")
        println()
        println("Concentration: ${character.concentration()}")
        println("Strength: ${character.strength()}")
        println("Agility: ${character.agility()}")
        println("Suppleness: ${character.suppleness()}")
        println("Quickness: ${character.quickness()}")
        println()
        println("Health: ${character.health()} / ${character.health(MAX)}")
        println("Stamina: ${character.stamina()} / ${character.stamina(MAX)}")
        println()
        println("Action left: ${character.action}")
        println("Stamina per dodge: ${character.neededStaminaToDodge}")
        println("Dodge frequency: ${character.dodgeFrequency()}/${character.dodgeFrequency(MAX)}")
        println("Dodge duration: ${character.dodgeDuration()}/${character.dodgeDuration(MAX)}")

        println()
        println("=================")
        println("Hand-to-hand:")
        println("=================")
        println()
        println("Damage:")
        Specification.values().forEach { println("$it = ${character.handToHandDamage(it)} (${character.handToHandCriticalDamage(it)})") }
        println()
        println("Stamina per attack: ${character.neededStaminaForHandToHandAttack}")
        println("Attack frequency: ${character.handToHandActionFrequency()}/${character.handToHandActionFrequency(MAX)}")
        println("Attack duration: ${character.handToHandAttackDuration()}/${character.handToHandAttackDuration(MAX)}")
        println("Defense duration: ${character.handToHandDefenseDuration()}/${character.handToHandDefenseDuration(MAX)}")

        if (character.primaryWeaponExist()) {
            println()
            println("=================")
            println(character.primaryWeapon().weaponType.toString().replace("_", " "))
            println("=================")
            println()
            println("Weight: ${character.primaryWeapon().weight}")
            println("Quality: ${character.primaryWeapon().quality}")
            println("Price: ${character.primaryWeapon().price}")
            println()
            println("Damage:")
            character.primaryWeapon().damage.forEach({ k, v -> println("$k = $v") })
            println()
            println("Stamina per use: ${character.neededStaminaForPrimaryWeaponAttack}")
            println("Attack frequency: ${character.primaryWeaponUsingFrequency()}/${character.primaryWeaponUsingFrequency(MAX)}")
            println("Attack duration: ${character.primaryWeaponAttackDuration()}/${character.primaryWeaponAttackDuration(MAX)}")
            println("Defense duration: ${character.primaryWeaponDefenseDuration()}/${character.primaryWeaponDefenseDuration(MAX)}")

            println()
            println("=================")
            println("Primary Weapon Damage:")
            println("=================")
            println()
            Specification.values().forEach { println("$it = ${character.primaryWeaponDamage(it)} (${character.primaryWeaponCriticalDamage(it)})") }
        }

        if (character.secondaryWeaponExist()) {
            println()
            println("=================")
            println(character.secondaryWeapon().weaponType.toString().replace("_", " "))
            println("=================")
            println()
            println("Weight: ${character.secondaryWeapon().weight}")
            println("Quality: ${character.secondaryWeapon().quality}")
            println("Price: ${character.secondaryWeapon().price}")
            println()
            println("Damage:")
            character.secondaryWeapon().damage.forEach({ k, v -> println("$k = $v") })
            println()
            println("Stamina per use: ${character.neededStaminaForSecondaryWeaponAttack}")
            println("Attack frequency: ${character.secondaryWeaponUsingFrequency()}/${character.secondaryWeaponUsingFrequency(MAX)}")
            println("Attack duration: ${character.secondaryWeaponAttackDuration()}/${character.secondaryWeaponAttackDuration(MAX)}")
            println("Defense duration: ${character.secondaryWeaponDefenseDuration()}/${character.secondaryWeaponDefenseDuration(MAX)}")

            println()
            println("=================")
            println("Secondary Weapon Damage:")
            println("=================")
            println()
            Specification.values().forEach { println("$it = ${character.secondaryWeaponDamage(it)} (${character.secondaryWeaponCriticalDamage(it)})") }
        }

        if (character.shieldExist()) {
            println()
            println("=================")
            println(character.shield().shieldType.toString().replace("_", " "))
            println("=================")
            println()
            println("Weight: ${character.shield().weight}")
            println("Quality: ${character.shield().quality}")
            println("Price: ${character.shield().price}")
            println()
            println("Stamina per use: ${character.neededStaminaForShieldUsing}")
            println("Using frequency: ${character.shieldUsingFrequency()}")
            println("Using duration: ${character.shieldUsingDuration()}")
            println()
            println("Defense:")
            character.shield().defense.forEach({ k, v -> println("$k = $v") })
        }

        if (character.has(HELM)) {
            println()
            println("=================")
            println(HELM.toString().replace("_", " "))
            println("=================")
            println()
            println("Weight: ${character.equipmentOf(HEAD).weight}")
            println("Quality: ${character.equipmentOf(HEAD).quality}")
            println("Price: ${character.equipmentOf(HEAD).price}")
            println()
            println("Defense:")
            (character.equipmentOf(HEAD) as Armor).defense.forEach({ k, v -> println("$k = $v") })
        }

        if (character.has(RIGHT_SHOULDER_ARMOR)) {
            println()
            println("=================")
            println(RIGHT_SHOULDER_ARMOR.toString().replace("_", " "))
            println("=================")
            println()
            println("Weight: ${character.equipmentOf(RIGHT_SHOULDER).weight}")
            println("Quality: ${character.equipmentOf(RIGHT_SHOULDER).quality}")
            println("Price: ${character.equipmentOf(RIGHT_SHOULDER).price}")
            println()
            println("Defense:")
            (character.equipmentOf(RIGHT_SHOULDER) as Armor).defense.forEach({ k, v -> println("$k = $v") })
        }

        if (character.has(LEFT_SHOULDER_ARMOR)) {
            println()
            println("=================")
            println(LEFT_SHOULDER_ARMOR.toString().replace("_", " "))
            println("=================")
            println()
            println("Weight: ${character.equipmentOf(LEFT_SHOULDER).weight}")
            println("Quality: ${character.equipmentOf(LEFT_SHOULDER).quality}")
            println("Price: ${character.equipmentOf(LEFT_SHOULDER).price}")
            println()
            println("Defense:")
            (character.equipmentOf(LEFT_SHOULDER) as Armor).defense.forEach({ k, v -> println("$k = $v") })
        }

        if (character.has(GLOVES)) {
            println()
            println("=================")
            println(GLOVES.toString().replace("_", " "))
            println("=================")
            println()
            println("Weight: ${character.equipmentOf(HANDS).weight}")
            println("Quality: ${character.equipmentOf(HANDS).quality}")
            println("Price: ${character.equipmentOf(HANDS).price}")
            println()
            println("Defense:")
            (character.equipmentOf(HANDS) as Armor).defense.forEach({ k, v -> println("$k = $v") })
        }

        if (character.has(BODY_ARMOR)) {
            println()
            println("=================")
            println(BODY_ARMOR.toString().replace("_", " "))
            println("=================")
            println()
            println("Weight: ${character.equipmentOf(BODY).weight}")
            println("Quality: ${character.equipmentOf(BODY).quality}")
            println("Price: ${character.equipmentOf(BODY).price}")
            println()
            println("Defense:")
            (character.equipmentOf(BODY) as Armor).defense.forEach { k, v -> println("$k = $v") }
        }

        if (character.has(RIGHT_LEG_ARMOR)) {
            println()
            println("=================")
            println(RIGHT_LEG_ARMOR.toString().replace("_", " "))
            println("=================")
            println()
            println("Weight: ${character.equipmentOf(RIGHT_LEG).weight}")
            println("Quality: ${character.equipmentOf(RIGHT_LEG).quality}")
            println("Price: ${character.equipmentOf(RIGHT_LEG).price}")
            println()
            println("Defense:")
            (character.equipmentOf(RIGHT_LEG) as Armor).defense.forEach({ k, v -> println("$k = $v") })
        }

        if (character.has(LEFT_LEG_ARMOR)) {
            println()
            println("=================")
            println(LEFT_LEG_ARMOR.toString().replace("_", " "))
            println("=================")
            println()
            println("Weight: ${character.equipmentOf(LEFT_LEG).weight}")
            println("Quality: ${character.equipmentOf(LEFT_LEG).quality}")
            println("Price: ${character.equipmentOf(LEFT_LEG).price}")
            println()
            println("Defense:")
            (character.equipmentOf(LEFT_LEG) as Armor).defense.forEach({ k, v -> println("$k = $v") })
        }

        if (character.has(BOOTS)) {
            println()
            println("=================")
            println("$BOOTS")
            println("=================")
            println()
            println("Weight: ${character.equipmentOf(FOOTS).weight}")
            println("Quality: ${character.equipmentOf(FOOTS).quality}")
            println("Price: ${character.equipmentOf(FOOTS).price}")
            println()
            println("Defense:")
            (character.equipmentOf(FOOTS) as Armor).defense.forEach({ k, v -> println("$k = $v") })
        }

        println()
        println("=================")
        println("Total Equipment:")
        println("=================")
        println()
        println("Weight: ${character.equipmentWeight}")
        println("Price: ${character.equipmentPrice}")
        println()
        println("Defense:")
        Specification.values().forEach { println("$it = ${character.totalDefense(it)}") }

        println("")
        println("=================")
        println("Skills:")
        println("=================")
        println()
        Skill.values().forEach { println("$it = ${character.skill(it)}") }
        println()
    }

//    fun printComparison(first: Character, second: Character) {
//        println()
//        println("=================")
//        println("Main Parameters:")
//        println("=================")
//        println()
//        println("Dodge chance: ")
//    }
}
